alternator/executor.cc

```diff
@@ -1882,18 +1882,149 @@ static std::string get_item_type_string(const rjson::value& v) {
+        throw api_error::validation(format("Malformed value object for filtering: {}", val));
+    }
+    const char* type = val.MemberBegin()->name.GetString();
+    rjson::value& v = val.MemberBegin()->value;
+    if (!members.empty()) {
+        if (type[0] != 'M' || !v.IsObject()) {
+            // If v is not an object (dictionary, map), none of the members
+            // can match. We return immediately because we assume that either
+            // members or indexes are set, but not both.
+            return false;
+        }
+        rjson::value newv = rjson::empty_object();
+        for (auto it = v.MemberBegin(); it != v.MemberEnd(); ++it) {
+            std::string attr = it->name.GetString();
+            auto h = members.find(attr);
+            if (h != members.end()) {
+                if (h->second) {
+                    // Only a part of this attribute is to be filtered, do it.
+                    if (h->second->filter(it->value)) {
+                        rjson::set_with_string_name(newv, attr, std::move(it->value));
+                    }
+                } else {
+                    // The entire attribute is to be kept
+                    rjson::set_with_string_name(newv, attr, std::move(it->value));
+                }
+            }
+        }
+        if (newv.MemberCount() == 0) {
+            return false;
+        }
+        v = newv;
+    } else if (!indexes.empty()) {
+        if (type[0] != 'L' || !v.IsArray()) {
+            return false;
+        }
+        rjson::value newv = rjson::empty_array();
+        const auto& a = v.GetArray();
+        for (unsigned i = 0; i < v.Size(); i++) {
+            auto h = indexes.find(i);
+            if (h != indexes.end()) {
+                if (h->second) {
+                    if (h->second->filter(a[i])) {
+                        rjson::push_back(newv, std::move(a[i]));
+                    }
+                } else {
+                    // The entire attribute is to be kept
+                    rjson::push_back(newv, std::move(a[i]));
+                }
+            }
+        }
+        if (newv.Size() == 0) {
+            return false;
+        }
+        v = newv;
+    }
+    return true;
+}
+void hierarchy_filter::add(const parsed::path& p) {
+    hierarchy_filter* h = this;
+    // new_level is false if the previous iteration of the loop below found
+    // its level in the hierarchy already existed, and true means that we
+    // just created this level now. We start with true because we assume
+    // the caller (calculate_attrs_to_get()) takes cares of not allowing
+    // the same top-level attribute twice.
+    bool new_level = true;
+    for (const auto& op : p.operators()) {
+        std::visit(overloaded_functor {
+            [&] (const std::string& member) {
+                if (!h->indexes.empty()) {
+                    throw api_error::validation(
+                        // TODO: print the conflicting paths
```

Comment:
nitpick: it would be easy to print at least the new path (`p`), right? And that gives much more debugging context already.

Reply:
Since I'm rewriting this part of the code anyway - to be reusable in UpdateExpression as well - and spending hours on avoiding code duplication, I guess I might as well spend a few minutes in doing this TODO and not be lazy :-)